Feature/center content

jkdufair opened the pull request:
Hello @rakanalh. This PR addresses https://github.com/emacs-dashboard/emacs-dashboard/issues/77  
I approached this by creating a macro that can insert arbtrary sections and returns the maximum line length for that section. Then `'dashboard-insert-startupify-lists` finds the maximum of the maximums and moves all items over accordingly. I initially looked at adjusting the margins and/or the gutter but neither looked very attractive.

Additionally, I'd be willing to be a maintainer. I've been looking for something fun to hack on for a while. Thanks for considering.

jkdufair commented:
Just realized I need to update the readme. Sorry about that.

jkdufair commented:
- Page break lines are now not "centered"
 - Documentation added about centering

rakanalh commented:
Hello @jkdufair,

Thank you so much for putting together a PR.

Before i review the code, please fix back the code indentation to be able to properly review the other parts of the code. I think we should introduce some lint to that because every contributor seems to have their own indentation style / settings.

jkdufair commented:
Sure thing. Do you have a style guide I can refer to?

rakanalh commented:
Not really... now i am considering it to be honest.

jkdufair commented:
Ok. I tried not to make any whitespace-only changes to existing functions. I'll revert any that I may have created inadvertently. Thank you.

jkdufair commented:
Ok sorry about that! I didn't realize I had those whitespace-only changes. Thanks for taking a look.

rakanalh commented on dashboard.el:
```diff
@@ -128,20 +133,28 @@
                     (add-to-list 'dashboard--section-starts (point))
-                    (funcall item-generator list-size)
+                    (setq max-line-length
+													(max max-line-length (or (funcall item-generator list-size) 0)))
```
Indentation is wrong here

jkdufair replied:
Fixed!

rakanalh commented on dashboard.el:
```diff
@@ -128,20 +133,28 @@
                     (add-to-list 'dashboard--section-starts (point))
-                    (funcall item-generator list-size)
+                    (setq max-line-length
+													(max max-line-length (or (funcall item-generator list-size) 0)))
```
Indentation is wrong here and the lines below

jkdufair replied:
Fixed!

rakanalh commented on dashboard.el:
```diff
@@ -128,20 +133,28 @@
-	(dashboard-mode)
-	(goto-char (point-min))))
+								dashboard-items)
+					(if dashboard-center-content
```
Please use `(when` for this instead if `(if (progn)`

jkdufair replied:
Fixed!

rakanalh commented on dashboard-widgets.el:
```diff
@@ -225,128 +225,100 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
+	`(let ((max-line-length 0))
+		 (when (car ,list)
+			 (mapc (lambda (el)
+							 (let ((widget nil))
```
Indentation is wrong here and the lines below

jkdufair replied:
Fixed!

rakanalh commented on dashboard-widgets.el:
```diff
@@ -225,128 +225,100 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
+						 ,list))
+		 max-line-length))
+
+(defmacro dashboard-insert-section (section-name list list-size shortcut action &rest widget-params)
```
Why did this become a macro rather than a function?

jkdufair replied:
I wanted to have each section insert return its longest line so I could do centering calculations. Each of the section and section list inserts was similar and I didn't want to just boilerplate it in there. So I refactored the section and section list inserts into a macro and replaced the repetitive sections with calls to the macro. This should also make it easy to create new sections in the future with less boilerplate code.

jkdufair replied:
Also it should make the intent of creating sections more clean and more declarative

rakanalh commented on dashboard-widgets.el:
```diff
@@ -225,128 +225,100 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
 (defun dashboard-insert-recents (list-size)
   "Add the list of LIST-SIZE items from recently edited files."
   (recentf-mode)
-  (when (dashboard-insert-recentf-list
```
Dashboard let's you configure widgets to include... this is why the `when` statement is there. Why was it removed?

jkdufair replied:
I believe this still works as intended? When I remove, i.e. `'recents` from `'dashboard-items`, the section no longer appears on the dashboard. Am I understanding correctly?

jkdufair replied:
Sorry. I should add there is a `when-let` in `dashboard-insert-section` that I believe accomplishes the same.

jkdufair replied:
Also, the linter seems to have had trouble with `when-let` so I just did a `let` and `if`. This still appears to allow the configurability of which widgets to include.

rakanalh commented on dashboard-widgets.el:
```diff
@@ -225,128 +225,100 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
+	 "m"
+	 `(lambda (&rest ignore) (bookmark-jump ,el))
+	 (let ((file (bookmark-get-filename el)))
+								 (if file
```
Indentation is wrong here and the lines below

jkdufair replied:
Fixed!

rakanalh commented on dashboard-widgets.el:
```diff
@@ -225,128 +225,100 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
+SHORTCUT is the keyboard shortcut used to access the section.
+WIDGET-PARAMS are passed to the \"widget-create\" function.
+Show EMPTY-LIST-TEXT if no items in list"
+	`(progn
```
Is there a specific reason of using the progn here?

jkdufair replied:
I am not aware of a way to return multiple forms to be executed in a `defmacro`. Can you advise a better method? I found nothing after a substantial amount of reading the manual, googling, etc.

rakanalh commented on dashboard-widgets.el:
```diff
@@ -225,128 +225,100 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
+	`(progn
+		 (dashboard-insert-heading ,section-name)
+		 (when-let ((max-line-length
+								 (dashboard-insert-section-list
```
Indentation is wrong here and the lines below

jkdufair replied:
Fixed!

JesusMtnez commented:
@jkdufair sorry for the conflict, just merge a PR that deletes `:follow-link "\C-m" which makes your macro change a conflict. :sweat_smile:

rakanalh commented on dashboard.el:
```diff
@@ -62,6 +62,11 @@
   "Settings that are used in the Dashboard"
   :group 'dashboard)
 
+(defcustom dashboard-center-content t
```
I would rather have the default value as nil (non-centered)... might need to update the documentation with this change.

jkdufair commented:
I agree. I don’t want to break backward compatibility @rakanalh. The more I think about it, the less I like putting the responsibility of returning `max-line-length` on each section. If I remove that requirement, I believe custom widgets will work just fine as-is. And people can use the new macros as well to easily develop more custom widgets. Should be a simple change and it will also make the macros simpler.

Sound ok?

jkdufair commented:
@jesusmtnez no problem. I’ll get it fixed shortly.

JesusMtnez commented:
@rakanalh I don't have any possible idea to that. 

@jkdufair I like you approach, but if it is possible to keep backwards compatibility, let's try it.

In case we do not find a way, I would suggest making an stable release before merging this, so next release is a major release.

Also, good documentation is important if this change is going to include breaking changes.

jkdufair commented:
@rakanalh @JesusMtnez I have done the following
- Removed the need for sections to return their maximum width. I calculate it, instead, as the sections are inserted
- Verified it is backward compatible by adding a custom section
- Set the default centering to `nil` and updated the documentation accordingly
- Merged upstream master and resolved conflicts.

Thank you for your assistance with this.

jkdufair commented:
Glad to help. Count me in as a regular contributor.
